src/accounts/nuls2.ts

```diff
@@ -132,10 +139,93 @@ export async function ImportAccountFromPrivateKey(
     privateKey: string,
     { chain_id = 1, prefix = "NULS" }: NULS2ImportConfig = { chain_id: 1, prefix: "NULS" },
 ): Promise<NULS2Account> {
-    const pub = privateKeyToPublicKey(Buffer.from(privateKey, "hex"));
+    const pub = secp256k1.publicKeyCreate(Buffer.from(privateKey, "hex"));
     const publicKey = Buffer.from(pub).toString("hex");
 
     const hash = publicKeyToHash(pub, { chain_id: chain_id });
     const address = addressFromHash(hash, prefix);
     return new NULS2Account(address, publicKey, privateKey);
 }
+
+/********************************
+ * NULS2 TOOLS METHODS
+ ********************************/
+
+/**
+ * Extract an address from a given hash.
+ *
+ * @param hash The hash containing the address.
+ * @param prefix The optional address prefix.
+ */
+export function addressFromHash(hash: Uint8Array, prefix?: string): string {
+    const address = bs58.encode(Buffer.concat([hash, Buffer.from([getXOR(hash)])]));
+
+    if (prefix) return prefix + String.fromCharCode(prefix.length + 96) + address;
+    return address;
+}
+
+/**
+ * Creates a XOR of an array.
+ *
+ * @param body The array to XOR.
+ */
+export function getXOR(body: Uint8Array): number {
+    return body.reduce((xor, i) => (xor ^= i), 0);
+}
+
+/**
+ * Creates a hash from a message.
+ *
+ * @param message The message used to create the hash.
+ * @param messagePrefix The optional message's hash prefix.
+ */
+export function magicHash(message: Buffer, messagePrefix?: string | Buffer): Buffer {
+    if (!messagePrefix) messagePrefix = "\u0018NULS Signed Message:\n";
+    if (!Buffer.isBuffer(messagePrefix)) messagePrefix = Buffer.from(messagePrefix);
+
+    let buffer = Buffer.allocUnsafe(messagePrefix.length + 6 + message.length);
+    let cursor = messagePrefix.copy(buffer, 0);
+    cursor += writeVarInt(message.length, buffer, cursor);
+    cursor += Buffer.from(message).copy(buffer, cursor);
+    buffer = buffer.slice(0, cursor);
+    return new shajs.sha256().update(buffer).digest();
+}
+
+/**
+ * Creates a hash from a user's public key.
+ *
+ * @param publicKey The public key used to create the hash.
+ * @param chain_id The optional chain id.
+ * @param address_type The optional address type.
+ */
+export function publicKeyToHash(
+    publicKey: Uint8Array,
+    { chain_id = 8964, address_type = 1 }: ChainNAddress = { chain_id: 8964, address_type: 1 },
+): Buffer {
+    const sha = new shajs.sha256().update(publicKey).digest();
+    const publicKeyHash = new RIPEMD160().update(sha).digest();
+    const output = Buffer.allocUnsafe(3);
+
+    output.writeInt16LE(chain_id, 0);
+    output.writeInt8(address_type, 2);
+    return Buffer.concat([output, publicKeyHash]);
+}
+
+export function writeVarInt(value: number, buf: Buffer, cursor: number): number {
```

Comment:
This function has no doc comment.
(I think we could get rid of this function that mutates an unsafe buffer, by rewriting the `magicHash` helper)

Reply:
I merged writeVarInt() and magicHash(), see: https://github.com/aleph-im/aleph-sdk-ts/pull/43/commits/e20d579ea6f8e641b4398c0bd720f6538dba3b17

Also, I thought it was not so useful to keep getXor as a tool function without Nuls1 in this context, so I removed it. What did you think about it?